### PR TITLE
Input Action Rules Inspector

### PR DESCRIPTION
**Pull request by jshowacre**
## Issues
The quaternion field isn't working for me, is there an alternative for this, or am I using it incorrectly?

## Preview 
inputactionrules

### CODE REVIEW

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityInputActionRulesProfileInspector.cs [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+            inputActionRulesDigital = serializedObject.FindProperty("inputActionRulesDigital");
+            inputActionRulesSingleAxis = serializedObject.FindProperty("inputActionRulesSingleAxis");
+            inputActionRulesDualAxis = serializedObject.FindProperty("inputActionRulesDualAxis");
+            inputActionRulesVectorAxis = serializedObject.FindProperty("inputActionRulesVectorAxis");
+            inputActionRulesQuaternionAxis = serializedObject.FindProperty("inputActionRulesQuaternionAxis");
+            inputActionRulesPoseAxis = serializedObject.FindProperty("inputActionRulesPoseAxis");
+
+            #region Interaction Constraint
+
+            digitalActionIds = MixedRealityManager.Instance.ActiveProfile.InputSystemProfile.InputActionsProfile.InputActions
```
Comment (jshowacre) [inline Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityInputActionRulesProfileInspector.cs:82]: Should we have these lambdas standardized?

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityInputActionRulesProfileInspector.cs [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+            serializedObject.Update();
+
+            EditorGUILayout.BeginVertical();
+            EditorGUILayout.LabelField(DigitalContent, EditorStyles.boldLabel);
+            if (GUILayout.Button(DigitalAddRuleContent, EditorStyles.miniButton))
+            {
+                var prop = CreateNewAction(inputActionRulesDigital);
+                if (prop != null)
+                {
+                    prop.boolValue = false;
```
Comment (jshowacre) [inline Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityInputActionRulesProfileInspector.cs]: Are setting these default values necessary?

**Comment by SimonDarksideJ** [comment]
Hi @Jarodshow thanks for pitching in, however this isn't what we were trying to aim for, as the focus is on easy usability.

Ultimately, the focus for configuring the "Rules" is on the InputAction not the Axis, so the screen configuration should be based on a list of Actions to configure, e.g.

* Detected Action
    -> Value (different view based on the Axis type of the selected action)
    -> Target Action

Then we can easily configure the rules by action, rather than by Axis.  We only store data by axis as it's easier from a data storage perspective.

> I did do a Mock screenshot for @StephenHodgson , which has gone now.  If you want it again, let me know and I'll craft it again

**Comment by jshowacre** [comment]
Thanks for the input! I'd like to try again, so could you send me that screenshot you were talking about?

**Comment by StephenHodgson** [comment]
Agreed, ideally this is also a single list of rules and not just showing what's there.

**Comment by SimonDarksideJ** [comment]
Brought my Paint.Net skills out again :P  A draft of what the list might look like, but feel free to tweak around that.
image
But basically a list, where you select an action, based on the action the correct criteria shows, with a drop down to select the output

**Comment by jshowacre** [comment]
Was this more of what you had in mind @StephenHodgson @SimonDarksideJ ?

inputactionrules3

**Comment by jshowacre** [comment]
Also, should the rule action have any constraints on it?

**Comment by StephenHodgson** [comment]
The constraints are built into defining the rule. If the constraints don't match then the equality check won't pass. You'll also need to filter out the actions for the user to see, based on the rule action.

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityInputActionRulesProfileInspector.cs [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+            for (int i = 0; i < inputActionRulesVectorAxis?.arraySize; i++)
+            { properties.Add(inputActionRulesVectorAxis.GetArrayElementAtIndex(i)); }
+            for (int i = 0; i < inputActionRulesQuaternionAxis?.arraySize; i++)
+            { properties.Add(inputActionRulesQuaternionAxis.GetArrayElementAtIndex(i)); }
+            for (int i = 0; i < inputActionRulesPoseAxis?.arraySize; i++)
+            { properties.Add(inputActionRulesPoseAxis.GetArrayElementAtIndex(i)); }
+            
+
+            actionProperties = properties
+                .OrderBy(property => property.FindPropertyRelative("baseAction").FindPropertyRelative("description").stringValue) // Sort by name or id?
```
Comment (jshowacre) [inline Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityInputActionRulesProfileInspector.cs]: Input on this?

**Comment by SimonDarksideJ** [comment]
The updated screenshot looks cool, a few points through:

- The fields should be as follows, to make it clearer to config when building rules:
    * Source Action
    * Criteria
    * Target Action
- Try and align the value fields if you can, so it looks a little more aligned :D 
- Might be good to enhance the Text / message box talking about Rules and their use above the section (see the other profiles for reference).
- (Stretch) would be good to group rules by source action.  To make it clear that several rules for a single Source Action are grouped together
- Is there validation to ensure that duplication is not allowed (Duplicate Source + Value + Target)

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityInputActionRulesProfileInspector.cs [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+                    prop.vector2Value = Vector2.zero;
+                    break;
+                case AxisType.ThreeDofPosition:
+                    prop.vector3Value = Vector3.zero;
+                    break;
+                case AxisType.ThreeDofRotation:
+                    prop.quaternionValue = Quaternion.identity;
+                    break;
+                case AxisType.SixDof:
+                    prop.boundsValue = new Bounds();
```
Comment (jshowacre) [inline Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityInputActionRulesProfileInspector.cs]: @StephenHodgson what value should i use for pose?

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityInputActionRulesProfileInspector.cs]: `MixedRealityPose.ZeroIdentity`

Comment (jshowacre) [reply Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityInputActionRulesProfileInspector.cs]: I mean the `prop.___value`, I couldnt find one that matched pose, unless there's an object one that I missed

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityInputActionRulesProfileInspector.cs]: There's not a built in one.

Comment (jshowacre) [reply Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityInputActionRulesProfileInspector.cs]: So how you go about setting its identity?

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityInputActionRulesProfileInspector.cs]: I think you have to get the vector3 and quaternion relative property values and set those.

**Comment by SimonDarksideJ** [comment]
Some review comments:
* Would be (great but not essential) of the "Groups" were in a collapsible section. Good grouping but could be tweaked
* Duplication checking can't just DELETE the duplicate, needs to warn the user somehow.  I was really confused before I realised what was happening when I intentionally created a default when it disappeared 
* The "Default" asset should be prefixed with "Default" like the other defaults 🗡 
* If "New" rules are going to appear at the top. Add a separator between "new" and the "Existing" items.  By default new has always gone to the bottom, so that is where users will look.  But I do like this pattern of configuring first before adding to the list.
* Linked to above, it's confusing that the list auto-sorts when you select the base action.  As the thing you started configuring jumps to "somewhere else" :S 
* Duplication check seems off, I created another rule "TeleportDirection" and also mapped that to "DPadDown" and it deleted :S 
* Boolean and Float value is still off

Will keep testing, but looking good

**Comment by jshowacre** [comment]
Sorry this took so long, I've been busy. 
The only requests left should be the separator and boolean float value alignment

**Comment by StephenHodgson** [comment]
Same same. I'll take a look later today after work.

**Comment by jshowacre** [comment]
Should the quaternion criteria be interpreted as Euler angles in the editor?

**Comment by StephenHodgson** [comment]
Quaternion

**Comment by SimonDarksideJ** [comment]
OK, review comments Mk3 :D

- The number next to the "Group", what is that meant to signify as the number doesn't currently relate to anything?  - Guessing it should be the number of rules in the group?. (Figured out it was the Action ID, but it's not clear.  Do we need this number?)

- If I create a duplicate, it throws a warning that there is a duplicate.  On clicking OK it saves anyway?  Probably need to check on any field edit for dupes, so you can't continue if there is one.

- Would be nice if the dupe check message also role you which Source Action has duplicates.

- Groups don't take effect until you dismiss and then select the profile again ? (is this a limitation)

- If we are grouping, then we might as well be uniform, so groups of 1 should be allowed? Thoughts @StephenHodgson 

Overall, a great improvement, well done @Jarodshow

**Comment by StephenHodgson** [comment]
Getting some errors on my end:
image

**Comment by StephenHodgson** [comment]
Sorry mate, I went ahead and did it.

**Comment by SimonDarksideJ** [comment]
I think what @StephenHodgson meant to say @Jarodshow is.  Thanks for the PR, it really helped to move the conversation along as to what is needed in this area and he's taken on board all your points and implemented them manually to speed things along.
(in his own way :P) We really appreciate your efforts in this area, so please keep contributing.

**Comment by StephenHodgson** [comment]
Yes, 1000% agreed.

**Pull request closed without merging**